File: Habitica/src/main/java/com/habitrpg/android/habitica/data/local/TaskLocalRepository.java
package com.habitrpg.android.habitica.data.local;

import com.habitrpg.android.habitica.models.tasks.ChecklistItem;
import com.habitrpg.android.habitica.models.tasks.RemindersItem;
import com.habitrpg.android.habitica.models.tasks.Task;
import com.habitrpg.android.habitica.models.tasks.TaskList;
import com.habitrpg.android.habitica.models.tasks.TaskTag;
import com.habitrpg.android.habitica.models.tasks.TasksOrder;

import java.util.List;

import io.realm.RealmResults;
import rx.Observable;

public interface TaskLocalRepository extends BaseLocalRepository {

    Observable<RealmResults<Task>> getTasks(String taskType, String userID);
    Observable<RealmResults<Task>> getTasks(String userId);

    void saveTasks(String userId, TasksOrder tasksOrder, TaskList tasks);

    void saveTask(Task task);

    void deleteTask(String taskID);

    Observable<Task> getTask(String taskId);
    Observable<Task> getTaskCopy(String taskId);

    void markTaskCompleted(String taskId, boolean isCompleted);

    void saveReminder(RemindersItem remindersItem);

    void swapTaskPosition(int firstPosition, int secondPosition);

    Observable<Task> getTaskAtPosition(int currentPosition);
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/data/local/TaskLocalRepository.kt
package com.habitrpg.android.habitica.data.local

import com.habitrpg.android.habitica.models.tasks.Task
import com.habitrpg.android.habitica.models.tasks.TaskList
import com.habitrpg.android.habitica.models.user.User
import com.habitrpg.shared.habitica.models.tasks.TaskType
import com.habitrpg.shared.habitica.models.tasks.TasksOrder
import io.reactivex.rxjava3.core.Maybe
import kotlinx.coroutines.flow.Flow

interface TaskLocalRepository : BaseLocalRepository {

    fun getTasks(taskType: TaskType, userID: String, includedGroupIDs: Array<String>): Flow<List<Task>>
    fun getTasks(userId: String): Flow<List<Task>>

    fun saveTasks(ownerID: String, tasksOrder: TasksOrder, tasks: TaskList)

    fun deleteTask(taskID: String)

    fun getTask(taskId: String): Flow<Task>
    fun getTaskCopy(taskId: String): Flow<Task>

    fun markTaskCompleted(taskId: String, isCompleted: Boolean)

    fun swapTaskPosition(firstPosition: Int, secondPosition: Int)

    fun getTaskAtPosition(taskType: String, position: Int): Flow<Task>

    fun updateIsdue(daily: TaskList): Maybe<TaskList>

    fun updateTaskPositions(taskOrder: List<String>)
    fun saveCompletedTodos(userId: String, tasks: MutableCollection<Task>)
    fun getErroredTasks(userID: String): Flow<List<Task>>
    fun getUser(userID: String): Flow<User>
    fun getTasksForChallenge(challengeID: String?, userID: String?): Flow<List<Task>>
}
